File: Runtime/Plugins/iOS/Source/NativeCallHandler.swift
import AffiseInternal

@objc
public class NativeCallHandler: NSObject {
    @objc(apiCallVoid:json:)
    public func apiCall(_ apiName: String, json: String) {
        let _ = apiCallAny(apiName, json: json)
    }

    @objc(apiCallString:json:)
    public func apiCall(_ apiName: String, json: String) -> String? {
        return apiCallToType(apiName, json: json)
    }

    @objc(apiCallBool:json:)
    public func apiCall(_ apiName: String, json: String) -> Bool {
        return apiCallToType(apiName, json: json) ?? false
    }

    private func apiCallAny(_ apiName: String, json: String) -> Any? {
        var result: AffiseResult = ApiResult()

        guard let api = AffiseApiMethod.from(apiName) else {
            print("AffiseApi [\(apiName)] error: method not found")
            return nil
        }

        do {
            if let dict = try JSONSerialization.jsonObject(with: Data(json.utf8), options: .mutableContainers) as? [String: Any?] {
                apiCall(api, data: dict, result: &result)
            }
        } catch {
            print("AffiseApi [\(apiName)] error: \(error.localizedDescription)")
            return nil
        }

        if (result as? ApiResult)?.isNotImplemented() == true {
            print("AffiseApi [\(apiName)] error: api not implemented")
            return nil
        }

        if let error = (result as? ApiResult)?.getError() {
            print("AffiseApi [\(apiName)] error: \(error)")
            return nil
        }

        return asNativeData(result)
    }
    
    private func asNativeData(_ result: AffiseResult) -> Any? {
        guard let result = result as? ApiResult else { return nil }
        guard let data = result.getResult() else { return nil }
                
        switch data {
        case is [String:Any?]:
            guard let data = data as? [String:Any?] else { return "{}" }
            return data.toJson()
        case is [Any?]:
            guard let data = data as? [Any?] else { return "[]" }
            return data.toJson()
        default:
            return data
        }
    }

    private func apiCallToType<T>(_ apiName: String, json: String) -> T? {
        guard let result: Any = apiCallAny(apiName, json: json) else {
            return nil
        }
        switch result {
        case is Bool:
            return result as? T
        case is String:
            return result as? T
        default:
            print("AffiseApi [\(apiName)] error: unsupported return type")
            return nil
        }
    }

    func apiCall(_ api: AffiseApiMethod, data: [String: Any?], result _: inout AffiseResult) {
        print("AffiseApi [\(api.method)] error: methods not implemented")
    }
}

File: Runtime/Plugins/unity.androidlib/src/main/java/com/affise/attribution/unity/NativeCallHandler.java
package com.affise.attribution.unity;


import com.affise.attribution.internal.AffiseApiMethod;
import com.affise.attribution.internal.callback.AffiseResult;
import com.affise.attribution.internal.utils.JSONObjectExtKt;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;
import java.util.Map;

public abstract class NativeCallHandler {

    public <T> T apiCall(String apiName, String json) throws Exception {
        ApiResult result = new ApiResult();
        JSONObject jsonObject = new JSONObject(json);

        apiCall(
                AffiseApiMethod.from(apiName),
                JSONObjectExtKt.toMap(jsonObject),
                result
        );

        if (result.isNotImplemented()) {
            throw new Exception(String.format("AffiseApi [%s] error: not implemented", apiName));
        }

        if (result.isError()) {
            throw new Exception(String.format("AffiseApi [%s] error: [%s]", apiName, result.getError()));
        }

        try {
            return asNativeData(result);
        } catch (Exception e) {
            throw new Exception(String.format("AffiseApi [%s] error: [%s]", apiName, e.getLocalizedMessage()));
        }
    }
    
    public int apiCallBool(String apiName, String json) throws Exception {
        Boolean value = apiCall(apiName, json);
        return value ? 1 : 0;
    }
    
    private <T> T asNativeData(ApiResult result) {
        Object data = result.getResult();

        if (data instanceof Map) {
            JSONObject json = new JSONObject((Map)data);
            return (T) json.toString();
        } else if (data instanceof List) {
            JSONArray json = new JSONArray((List)data);
            return (T) json.toString();
        }

        return (T) data;
    }

    public void apiCallVoid(String apiName, String json) throws Exception {
        apiCall(apiName, json);
    }

    public abstract void apiCall(AffiseApiMethod api, Map<String, ?> data, AffiseResult result);
}
